### PR TITLE
Service state recovery

### PR DESCRIPTION
**Pull request by roysc**
Implements https://github.com/vulcanize/ipld-eth-state-snapshot/issues/17

Also includes corrections for edge cases, cleans up the logic some, and expands tests.
Depends on https://github.com/vulcanize/go-eth-state-node-iterator/pull/10, will update go.mod after that is merged.

### CODE REVIEW

File: pkg/snapshot/file/publisher_test.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+	headerID := fixt.Block1_Header.Hash().String()
+	test.NoError(t, pub.PublishStateNode(&fixt.Block1_StateNode0, headerID, tx))
+
+	test.NoError(t, tx.Commit())
+	return pub
+}
+
+// verify that we can parse the csvs
+// TODO check actual data
```
Comment (i-norden) [inline pkg/snapshot/file/publisher_test.go:48]: This is a key TODO, and something we need to do before deploying this.

I think the first approach we should take here is to run this new state of the snapshot tool at some blockheight that has state and storage trie nodes but is early enough to finish in a semi-reasonable amount of time (e.g. block 100,000). Have it write that out to csv files, then load that into Postgres, and use https://github.com/vulcanize/ipld-eth-db-validator to confirm everything is in Postgres.

This should be an automated integration test but for that we need to either port around a levelDB snapshot at/near that height or setup our LevelDB RPC server and start using that in these types of tests (ideal case).

For a first pass, and as a requirement to get this merged, you can  setup and run that test manually.

File: pkg/snapshot/tracker.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	"fmt"
+	"os"
+
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/trie"
+
+	iter "github.com/vulcanize/go-eth-state-node-iterator"
+)
+
+type trackedIter struct {
```
Comment (i-norden) [inline pkg/snapshot/tracker.go:14]: Do we have a sense of how much overhead this iteration tracking adds?

File: go.mod [hunk 3]
```diff
@@ -42,4 +43,6 @@ require (
 	lukechampine.com/blake3 v1.1.7 // indirect
 )
 
-replace github.com/ethereum/go-ethereum v1.10.15 => github.com/vulcanize/go-ethereum v1.10.15-statediff-2.0.0
+replace github.com/ethereum/go-ethereum v1.10.15 => github.com/vulcanize/go-ethereum v1.10.15-statediff-3.0.1
+
+replace github.com/vulcanize/go-eth-state-node-iterator v1.0.0 => ../state-node-iterator
```
Comment (i-norden) [inline go.mod:48]: The tagged release for this is ready.

**Review by i-norden — CHANGES_REQUESTED** [review]

File: pkg/snapshot/tracker.go (hunk 2 above)
Comment (roysc) [reply pkg/snapshot/tracker.go:14]: It seems pretty negligible. I didn't see any consistent slowdown with 8, 16, 32 threads

File: go.mod (hunk 3 above)
Comment (roysc) [reply go.mod:48]: updated

**Review by i-norden — APPROVED** [review]
LGTM! Thanks! Excited to try out all the changes.
